KAFKA-8403: Suppress needs a Materialized variant

dongjinleekr opened the pull request:
This is a completely new implemetation from the previous PR.

- [KAFKA-8403: Suppress needs a Materialized variant](https://issues.apache.org/jira/browse/KAFKA-8403)
- [KIP-508: Make Suppression State Queriable](https://cwiki.apache.org/confluence/display/KAFKA/KIP-508%3A+Make+Suppression+State+Queriable)

Here are some guidelines:

- 9bd279f ~ 37060b6: Refactor the `TimeOrderedKeyValueStore` into new interfaces, with separating `Serdes` from BytesStore. (see `MeteredSuppressionBuffer`)
- 798af5c ~ 0a684e0: Implement the `KTable#suppress` variant and make it queriable.
- 8f2e2c8  ~ 3387e27: Make `SuppressionBuffer` accessable by the users.
- 821014f: Implement a testing functionality.

Please have a look when you are free. Thanks in advance! :smile:

cc/ @mjsax @bbejeck @vvcephei @guozhangwang 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr commented:
Retest this please.

guozhangwang commented:
Wow that's a really large PR :) cc @vvcephei for a quick look.

vvcephei commented:
Hey @dongjinleekr , from the discussion on KIP-508, it seems this is not the approach we're going to take. Is that your impression as well?

dongjinleekr commented:
@vvcephei Sorry for being late, I just got out from my last project; I will have a look.

dongjinleekr commented:
@vvcephei Here is the fix. The implementation you saw was the draft before the discussion - Now it is updated, with rebasing onto the latest trunk. Don't hesitate to talk to me if you think we need additional tests like a queriable store.

Thanks again for your detailed review! :smile:

dongjinleekr commented:
Rebased onto the latest trunk. cc/ @vvcephei

dongjinleekr commented:
Hello. I updated the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-508%3A+Make+Suppression+State+Queriable) and rebased the PR onto the latest trunk. Please have a look when you are free.

cc/ @vvcephei @mjsax

dongjinleekr commented:
Rebased onto the latest trunk.

@vvcephei Could you have a look? It this implementation is okay, then I will open the vote.

cc/ @mjsax @bbejeck

bbejeck commented:
>@vvcephei Could you have a look? It this implementation is okay, then I will open the vote.

@dongjinleekr you don't have the implementation completed to call for a vote.  Maybe you can ping the discussion thread one last time to see if everything is settled and if its completed go ahead and call for a vote.

bbejeck commented:
test this please

dongjinleekr commented:
Rebased onto the latest trunk.

dongjinleekr committed:
Separate Serde, metric related fields from InMemoryTimeOrderedKeyValueBuffer

1. Separate InMemoryTimeOrderedKeyValueBuffer.Builder to TimedOrderedKeyValueBufferBuilder
2. Add MeteredTimeOrderedKeyValueBuffer, which handles Serde and metric related operations.
3. InMemoryTimeOrderedKeyValueBuffer<K, V> is now InMemoryTimeOrderedKeyValueBuffer, implementing TimedOrderedKeyValueBufferBuilder<Bytes, byte[]>
4. TimeOrderedKeyValueBuffer#evictWhile now returns int, instead of nothing; for supporting callback in MeteredTimeOrderedKeyValueBuffer#evictWhile.

dongjinleekr committed:
Make TimeOrderedKeyValueBuffer support ReadOnlyKeyValueStore operations

1. TimeOrderedKeyValueBuffer<K, V> now extends ReadOnlyKeyValueStore<K, V>

  New methods added: get, range, reverseRange, all, reverseAll, approximateNumEntries (default to numRecords().)

2. Change InMemoryTimeOrderedKeyValueBuffer#index's type from Map to NavigableMap

  Now supports range queries.

3. Add MeteredTimeOrderedKeyValueBuffer#valueSerde

  To support get operation.

4. Add InMemoryTimeOrderedKeyValueBuffer.InMemoryKeyValueIterator, MeteredTimeOrderedKeyValueBuffer.MeteredKeyValueIterator

  To support iterations.

5. MINOR: Change InMemoryTimeOrderedKeyValueBuffer#sortedMap's type from TreeMap to SortedMap

  Using interface, instead of concrete implementation.

dongjinleekr committed:
Add Suppressed#{enableQuery,isQueryEnabled}

1. Add FinalResultsSuppressionBuilder#{enableQuery,isQueryEnabled}
2. Add SuppressedInternal#{enableQuery,isQueryEnabled}
3. Update SuppressedTest.{java, scala} accordingly.

dongjinleekr committed:
Make Suppression state queriable, with Suppressed#{enableQuery, isQueryEnabled}

1. Make Suppression state queriable in KTableImp.java and Add documentation in KTable.{java,scala}.

2. Validate Interactive Query should find TimeOrderedKeyValueBuffers: StreamThreadStateStoreProviderTest#shouldFindTimeOrderedKeyValueBuffers

3. TRIVIAL: Add nullity check in KTableImpl#suppress(Suppressed)

4. TRIVIAL: Remove unused method, QueryableStateIntegrationTest.ProducerRunnable#shutdown

dongjinleekr commented:
@vvcephei Here is the update. it now consists of 4 stages:

1. Separating instantiating/serde/metric related operations from `InMemoryTimeOrderedKeyValueBuffer`.

    For `InMemoryTimeOrderedKeyValueBuffer` needs record collector and its changelog topic to implement `flush` operation, it has had a monolithic architecture dislike to the other `StateStore`s. And this approach is also making it difficult to support iterator.

    As a laying the groundwork, `TimeOrderedKeyValueBufferBuilder`, `MeteredTimeOrderedKeyValueBuffer` were separated into independent classes. `InMemoryTimeOrderedKeyValueBuffer` now concentrates on `Bytes`, `byte[]` operations without any serdes/metric operations.

2. By changing `InMemoryTimeOrderedKeyValueBuffer#index`'s type into `NavigatableMap`, it now supports `all`, `range` operations.

3. KtableImpl takes queriableStoreName into account.
4. KtableImpl now makes queriable `TimeOrderedKeyValueBuffer` instance.

Please have a look when you are free.
